### PR TITLE
RyuJIT: Don't insert GT_COMMA for x*2->x+x opt before optHoistLoopCode phase

### LINKED ISSUE #37802: [Perf -40%] Benchstone.BenchF.Simpsn.Test
### Run Information

Architecture | x64
-- | --
OS | Windows 10.0.18362
Changes | [diff](https://github.com/dotnet/installer/compare/a4a49f07fbf986d496ef57521e76b22caaf9bfb3...ccd701fe23f236755650af3ce3d9ab6bdd8a4af9)

### Regressions in Benchstone.BenchF.Simpsn

Benchmark | Baseline | Test | Test/Base | Modality | Baseline Outlier
-- | -- | -- | -- | -- | --
Test | 164.06 ms | 230.26 ms | 1.40 |  | False

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/06_11_2020/Windows%2010.0.18362_x64/Benchstone.BenchF.Simpsn.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'Benchstone.BenchF.Simpsn*';
```
<details>

#### Benchstone.BenchF.Simpsn.Test

```log
[162358394.287 ; 177057262.379) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[177057262.379 ; 185020170.713) | @
[185020170.713 ; 199719038.804) | 
[199719038.804 ; 214417906.896) | 
[214417906.896 ; 223408815.954) | 
[223408815.954 ; 239552934.046) | @@@@@@@
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/37802

Try to avoid GT_COMMA insertion before `optHoistLoopCode` phase to let it hoist such expressions first. A minimal repro from https://github.com/dotnet/runtime/issues/37802:
```csharp
static void Test(double x)
{
    for (int i = 0; i < 1000; i++)
    {
        Consume(Math.Exp(-x * 2)); // Math.Exp(-x * 2) should be hoisted from the loop
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Consume(double x){ }
```

Current codegen:
```asm
; Method Program:Test(double)
G_M55510_IG01:
       56                   push     rsi
       4883EC30             sub      rsp, 48
       C5F877               vzeroupper 
       C5F829742420         vmovaps  qword ptr [rsp+20H], xmm6
						;; bbWeight=1    PerfScore 5.25

G_M55510_IG02:
       33F6                 xor      esi, esi
       C5F828F0             vmovaps  xmm6, xmm0
       C5FB100D3C000000     vmovsd   xmm1, qword ptr [reloc @RWD08]
       C5C857F1             vxorps   xmm6, xmm1
						;; bbWeight=1    PerfScore 2.83

G_M55510_IG03:
       C5F828C6             vmovaps  xmm0, xmm6
       C5FB58C0             vaddsd   xmm0, xmm0, xmm0
       E85323B45F           call     System.Math:Exp(double):double
       E88699FEFF           call     Program:Consume(double)
       FFC6                 inc      esi
       81FEE8030000         cmp      esi, 0x3E8
       7CE4                 jl       SHORT G_M55510_IG03
						;; bbWeight=4    PerfScore 27.00

G_M55510_IG04:
       C5F828742420         vmovaps  xmm6, qword ptr [rsp+20H]
       4883C430             add      rsp, 48
       5E                   pop      rsi
       C3                   ret      
						;; bbWeight=1    PerfScore 5.75
RWD00  dq	0000000000000000h
RWD08  dq	8000000000000000h
; Total bytes of code: 72
```

New codegen:
```asm
; Method Program:Test(double)
G_M55510_IG01:
       56                   push     rsi
       4883EC30             sub      rsp, 48
       C5F877               vzeroupper 
       C5F829742420         vmovaps  qword ptr [rsp+20H], xmm6
						;; bbWeight=1    PerfScore 5.25

G_M55510_IG02:
       33F6                 xor      esi, esi
       C5FB100D40000000     vmovsd   xmm1, qword ptr [reloc @RWD08]
       C5F857C1             vxorps   xmm0, xmm1
       C5FB58C0             vaddsd   xmm0, xmm0, xmm0
       E85B23B75F           call     System.Math:Exp(double):double
       C5F828F0             vmovaps  xmm6, xmm0
						;; bbWeight=1    PerfScore 6.83

G_M55510_IG03:
       C5F828C6             vmovaps  xmm0, xmm6
       E88699FEFF           call     Program:Consume(double)
       FFC6                 inc      esi
       81FEE8030000         cmp      esi, 0x3E8
       7CED                 jl       SHORT G_M55510_IG03
						;; bbWeight=4    PerfScore 11.00

G_M55510_IG04:
       C5F828742420         vmovaps  xmm6, qword ptr [rsp+20H]
       4883C430             add      rsp, 48
       5E                   pop      rsi
       C3                   ret      
						;; bbWeight=1    PerfScore 5.75
RWD00  dq	0000000000000000h
RWD08  dq	8000000000000000h
; Total bytes of code: 72
```
Diff: https://www.diffchecker.com/y6dL37Qi

/cc @AndyAyersMS

### CODE REVIEW

**Comment by EgorBo** [comment]
"Cast" seems also is not hoisted:
```csharp
public class Program
{
    static void Test(object o)
    {
        for (int i = 0; i < 1000; i++) 
            Consume((Program) o); // cast o to Program is not hoisted from the loop
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Consume(Program p) {}
}
```

**Comment by EgorBo** [comment]
GT_MOD by constant also inserts a GT_COMMA in the global morph phase and prevents the following code to be optimized;
```csharp
static void Test(int x)
{
    for (int i = 0; i < 1000; i++) 
        Consume(-x % 10); // -x % 10 is not hoisted out of the loop
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Consume(int x) {}
```

**Review by AndyAyersMS — COMMENTED** [review]
Cast can cause exceptions, so hoisting is problematic.

How hard is it to fix the GT_MOD case?

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -13147,12 +13147,19 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
-                        op2  = op1->OperIsLeaf() ? gtCloneExpr(op1) : fgMakeMultiUse(&tree->AsOp()->gtOp1);
-                        op1  = tree->AsOp()->gtOp1;
-                        oper = GT_ADD;
-                        tree = gtNewOperNode(oper, tree->TypeGet(), op1, op2);
-                        INDEBUG(tree->gtDebugFlags |= GTF_DEBUG_NODE_MORPHED);
+                        bool needsComma = !op1->OperIsLeaf() && !op1->IsLocal();
+                        // if op1 is not a leaf/local we have to introduce a temp via GT_COMMA.
+                        // Unfortunately, it's not optHoistLoopCode-friendly yet so let's do it later
+                        // in the Rationalization phase.
+                        if (!needsComma || (mostRecentlyActivePhase > PHASE_HOIST_LOOP_CODE))
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/src/jit/morph.cpp]: @AndyAyersMS it does for the snippet in the description. that MUL is visited from `RewriteIntrinsicAsUserCall` during the rationalization phase (`Exp` is not intrinsic so we convert it back to a normal call and call `fgMorphArgs` for the arg). However, it won't be visited for:

```csharp
static void Test(double x)
{
    for (int i = 0; i < 1000; i++)
    {
        Consume(-x * 2); // Math.Exp(-x * 2) should be hoisted from the loop
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Consume(double x) { }
```
the expression will be hoisted but jit won't re-visit that node during rationalization.

**Review by AndyAyersMS — APPROVED** [review]
LGTM. Thanks for fixing this!

**Pull request merged**